### PR TITLE
Fix `fcollect` to respect object identity instead of `==`; document order

### LINKED ISSUE #16: `fcollect` won't keep repeated arrays that are `==` but distinct
See <https://github.com/beacon-biosignals/LegolasFlux.jl/pull/4#issuecomment-875546970> and the alternate implementation
```julia
function fcollect2(x; output = [], cache = Base.IdSet(), exclude = v -> false)
    x in cache && return output
    if !exclude(x)
      push!(cache, x)
      push!(output, x)
      foreach(y -> fcollect2(y; cache = cache, output=output, exclude = exclude), Functors.children(x))
    end
    return output
end
```

### PR DESCRIPTION
**Pull request by ericphanson**
Fixes #16. I checked the test I added failed on master:
```julia
julia> m1 = [1, 2, 3]
3-element Vector{Int64}:
 1
 2
 3

julia> m2 = [1, 2, 3]
3-element Vector{Int64}:
 1
 2
 3

julia> m3 = Foo(m1, m2)
Foo([1, 2, 3], [1, 2, 3])

julia> fcollect(m3)
2-element Vector{Any}:
 Foo([1, 2, 3], [1, 2, 3])
 [1, 2, 3]

julia> @test all(fcollect(m3) .=== [m3, m1, m2])
Error During Test at REPL[8]:1
  Test threw exception
  Expression: all(fcollect(m3) .=== [m3, m1, m2])
  DimensionMismatch("arrays could not be broadcast to a common size; got a dimension with lengths 2 and 3")
```

### CODE REVIEW

**Review by darsnack — APPROVED** [review]
Looks correct to me!

File: src/functor.jl [hunk 1]
```diff
@@ -192,11 +192,12 @@ julia> fcollect(m, exclude = v -> Functors.isleaf(v))
...
 """
-function fcollect(x; cache = [], exclude = v -> false)
-  x in cache && return cache
-  if !exclude(x)
-    push!(cache, x)
-    foreach(y -> fcollect(y; cache = cache, exclude = exclude), children(x))
-  end
-  return cache
+function fcollect(x; output = [], cache = Base.IdSet(), exclude = v -> false)
+    x in cache && return output
```
Comment (ericphanson) [inline src/functor.jl:196]: I realized it wasn't totally obvious why the code was done this way and also that LegolasFlux will depend on the order (for the same model) not changing in non-breaking releases of Functors.jl, so maybe this comment will help communicate that:
```suggestion
    # note: we don't have an `OrderedIdSet`, so we use an `IdSet` for the cache
    # (to ensure we get exactly 1 copy of each distinct array), and a usual `Vector`
    # for the results, to preserve traversal order (important downstream!).
    x in cache && return output
```
Note the tests already test the order, so hopefully that will be another sign to not break it 😄.

**Comment by DhairyaLGandhi** [comment]
If we intend to preserve the order, its worth a mention in the docs.

**Comment by ericphanson** [comment]
Ok, I wrote up the order defined by the implementation in the docstring. We could also not say exactly what the order is but just say it won't change for any inputs in a non-breaking release-- would that be preferred?

**Review by darsnack — APPROVED** [review]

**Comment by darsnack** [comment]
I think the current explanation is more precise and understandable.

**Pull request merged**